File: src/peer_ims.py
# This code is distributed under the GNU General Public License (see
# THE_GENERAL_GNU_PUBLIC_LICENSE.txt for extending this information).
# Copyright (C) 2014, the P2PSP team.
# http://www.p2psp.org

# The P2PSP.org project has been supported by the Junta de Andalucia
# through the Proyecto Motriz "Codificacion de Video Escalable y su
# Streaming sobre Internet" (P10-TIC-6548).

# {{{ Imports

from __future__ import print_function
import threading
import sys
import socket
import struct
from color import Color
import common
import time
from _print_ import _print_

# }}}

# IMS: Ip Multicasting Set of rules
class Peer_IMS(threading.Thread):
    # {{{

    # {{{ Class "constants"

    PLAYER_PORT = 9999          # Port used to serve the player.
    SPLITTER_ADDR = "127.0.0.1" # Address of the splitter.
    SPLITTER_PORT = 4552        # Port of the splitter.
    PORT = 0                    # TCP->UDP port used to communicate.
    USE_LOCALHOST = False       # Use localhost instead the IP of the addapter

    # }}}

    def __new__(typ, *args, **kwargs):
        # {{{

        if len(args) == 1 and isinstance(args[0], Peer_IMS):
            # Parameter is a peer instance; extending its class instead of nesting:
            instance = args[0]
            instance.__class__ = typ
            return instance
        else:
            # Use default object creation
            return object.__new__(typ, *args, **kwargs)

        # }}}

    def __init__(self):
        # {{{

        threading.Thread.__init__(self)
        sys.stdout.write(Color.yellow)
        _print_("Peer IMS")
        sys.stdout.write(Color.none)

        threading.Thread.__init__(self)

        _print_("Player port =", self.PLAYER_PORT)
        _print_("Splitter =", self.SPLITTER_ADDR)
        _print_("(Peer) port =", self.PORT)

        # }}}

    def wait_for_the_player(self):
        # {{{ Setup "player_socket" and wait for the player

        self.player_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            # In Windows systems this call doesn't work!
            self.player_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        except Exception as e:
            _print_ (e)
            pass
        self.player_socket.bind(('', self.PLAYER_PORT))
        self.player_socket.listen(0)
        _print_("Waiting for the player at", self.player_socket.getsockname())
        self.player_socket = self.player_socket.accept()[0]
        #self.player_socket.setblocking(0)
        _print_("The player is", self.player_socket.getpeername())
        
        # }}}

    def connect_to_the_splitter(self):
        # {{{ Setup "splitter" and "splitter_socket"

        # Nota: Ahora no reconvertimos de TCP a UDP!
        
        self.splitter_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.splitter = (self.SPLITTER_ADDR, self.SPLITTER_PORT)
        print("use_localhost=", self.USE_LOCALHOST)
        if self.USE_LOCALHOST:
            my_ip = '0.0.0.0' # Or '127.0.0.1'
            #my_ip = '127.0.0.1'
        else:
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            s.connect(self.splitter)
            #my_ip = socket.gethostbyname(socket.gethostname())
            my_ip = s.getsockname()[0]
            s.close()
        _print_("Connecting to the splitter at", self.splitter, "from", my_ip)
        if self.PORT != 0:
            try:
                self.splitter_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            except Exception as e:
                print (e)
                pass
            sys.stdout.write(Color.purple)
            _print_("I'm using port the port", self.PORT)
            sys.stdout.write(Color.none)
            self.splitter_socket.bind((my_ip, self.PORT))
        else:
            self.splitter_socket.bind((my_ip, 0))
        try:
            self.splitter_socket.connect(self.splitter)
        except Exception as e:
            print(e)
            sys.exit("Sorry. Can't connect to the splitter at " + str(self.splitter))
        _print_("Connected to the splitter at", self.splitter)
        
        # }}}

    def disconnect_from_the_splitter(self):
        # {{{

        self.splitter_socket.close()

        # }}}
        
    def receive_the_mcast_endpoint(self):
        # {{{
        message = self.splitter_socket.recv(struct.calcsize("4sH"))
        addr, port = struct.unpack("4sH", message)
        self.mcast_addr = socket.inet_ntoa(addr)
        self.mcast_port = socket.ntohs(port)
        mcast_endpoint = (self.mcast_addr, self.mcast_port)
        if __debug__:
            print("mcast_endpoint =", mcast_endpoint)
    
        # }}}

    def receive_the_header(self):
        # {{{

        header_size_in_bytes = self.header_size_in_chunks * self.chunk_size
        received = 0
        data = ""
        while received < header_size_in_bytes:
            data = self.splitter_socket.recv(header_size_in_bytes - received)
            received += len(data)
            _print_("Percentage of header received = {:.2%}".format((1.0*received)/header_size_in_bytes))
            try:
                self.player_socket.sendall(data)
            except Exception as e:
                print (e)
                print ("error sending data to the player")
                print ("len(data) =", len(data))
                time.sleep(1)
            _print_("received bytes:", received, "\r", end="")

        _print_("Received", received, "bytes of header")

        # }}}

    def receive_the_chunk_size(self):
        # {{{

        # The size of the chunk in bytes.
        message = self.splitter_socket.recv(struct.calcsize("H"))
        chunk_size = struct.unpack("H", message)[0]
        self.chunk_size = socket.ntohs(chunk_size)
        _print_("chunk_size (bytes) =", self.chunk_size)
        self.message_format = "H" + str(self.chunk_size) + "s"
        if __debug__:
            print ("message_format = ", self.message_format)
        
        # }}}

    def receive_the_header_size(self):
        # {{{

        message = self.splitter_socket.recv(struct.calcsize("H"))
        value = struct.unpack("H", message)[0]
        self.header_size_in_chunks = socket.ntohs(value)
        _print_("header_size (in chunks) =", self.header_size_in_chunks)

        # }}}

    def receive_the_buffer_size(self):
        # {{{

        message = self.splitter_socket.recv(struct.calcsize("H"))
        buffer_size = struct.unpack("H", message)[0]
        self.buffer_size = socket.ntohs(buffer_size)
        _print_("buffer_size =", self.buffer_size)

        # }}}

    def listen_to_the_team(self):
        # {{{ Create "team_socket" (UDP) for using the multicast channel

        self.team_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
        try:
            self.team_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        except Exception as e:
            print (e)
            pass

        try:
            self.team_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
        except Exception as e:
            print (e)
            pass

        self.team_socket.bind(('', self.mcast_port))
#        self.team_socket.bind(('', self.SPLITTER_SOCKET.getsockname()[PORT]))

        mreq = struct.pack("4sl", socket.inet_aton(self.mcast_addr), socket.INADDR_ANY)
        self.team_socket.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)
        _print_("Listening to the mcast_channel =", (self.mcast_addr, self.mcast_port))
        
        # This is the maximum time the peer will wait for a chunk
        # (from the splitter).
        self.team_socket.settimeout(1)
        if __debug__:
            print(self.team_socket.getsockname(), "\b.timeout = 1")
        
        # }}}

    def unpack_message(self, message):
        # {{{

        chunk_number, chunk = struct.unpack(self.message_format, message)
        chunk_number = socket.ntohs(chunk_number)

        return chunk_number, chunk

        # }}}
        
    def receive_the_next_message(self):
        # {{{

        if __debug__:
            print ("Waiting for a chunk at {} ...".format(self.team_socket.getsockname()))

        message, sender = self.team_socket.recvfrom(struct.calcsize(self.message_format))
        self.recvfrom_counter += 1

        # {{{ debug
        if __debug__:
            print (Color.cyan, "Received a message from", sender, \
                "of length", len(message), Color.none)
        # }}}

        return message, sender

        # }}}
        
    def process_next_message(self):
        # {{{
        try:
            # {{{ Receive a chunk

            message, sender = self.receive_the_next_message()
            chunk_number, chunk = self.unpack_message(message)
            
            self.chunks[chunk_number % self.buffer_size] = chunk
            self.received_flag[chunk_number % self.buffer_size] = True
            self.received_counter += 1

            return chunk_number

            # }}}
        except socket.timeout:
            return -2

        # }}}

    def buffer_data(self):
        # {{{ Buffering

        # {{{ The peer dies if the player disconnects.
        # }}}
        self.player_alive = True

        # {{{ The last chunk sent to the player.
        # }}}
        self.played_chunk = 0

        # {{{ Counts the number of executions of the recvfrom()
        # function.
        # }}}
        self.recvfrom_counter = 0

        # {{{ Label the chunks in the buffer as "received" or "not
        # received".
        # }}}
        self.received_flag = []

        # {{{ The buffer of chunks is a structure that is used to delay
        # the playback of the chunks in order to accommodate the
        # network jittter. Two components are needed: (1) the "chunks"
        # buffer that stores the received chunks and (2) the
        # "received" buffer that stores if a chunk has been received
        # or not. Notice that each peer can use a different
        # buffer_size: the smaller the buffer size, the lower start-up
        # time, the higher chunk-loss ratio. However, for the sake of
        # simpliticy, all peers will use the same buffer size.
        
        self.chunks = [""]*self.buffer_size
        self.received_flag = [False]*self.buffer_size
        self.received_counter = 0

        # }}}

        #  Wall time (execution time plus waiting time).
        start_time = time.time()

        # We will send a chunk to the player when a new chunk is
        # received. Besides, those slots in the buffer that have not been
        # filled by a new chunk will not be send to the player. Moreover,
        # chunks can be delayed an unknown time. This means that (due to the
        # jitter) after chunk X, the chunk X+Y can be received (instead of the
        # chunk X+1). Alike, the chunk X-Y could follow the chunk X. Because
        # we implement the buffer as a circular queue, in order to minimize
        # the probability of a delayed chunk overwrites a new chunk that is
        # waiting for traveling the player, we wil fill only the half of the
        # circular queue.

        _print_(self.team_socket.getsockname(), "\b: buffering = 000.00%")
        sys.stdout.flush()

        # First chunk to be sent to the player.  The
        # process_next_message() procedure returns the chunk number if
        # a packet has been received or -2 if a time-out exception has
        # been arised.
        chunk_number = self.process_next_message()
        while chunk_number < 0:
            chunk_number = self.process_next_message()
            print(chunk_number)
        self.played_chunk = chunk_number
        _print_("First chunk to play", self.played_chunk)
        _print_(self.team_socket.getsockname(), "\b: buffering (\b", repr(100.0/self.buffer_size).rjust(4))

        # Now, fill up to the half of the buffer.
        for x in range(int(self.buffer_size/2)):
            _print_("{:.2%}\r".format((1.0*x)/(self.buffer_size/2)), end='')
            #print("!", end='')
            sys.stdout.flush()
            while self.process_next_message() < 0:
                pass

        print()
        latency = time.time() - start_time
        _print_('latency =', latency, 'seconds')
        _print_("buffering done.")
        sys.stdout.flush()

        # }}}

    # Tiene pinta de que los tres siguientes metodos pueden
    # simplificarse...

    def find_next_chunk(self):
        # {{{
        #print (".")
        #counter = 0
        chunk_number = (self.played_chunk + 1) % common.MAX_CHUNK_NUMBER
        while not self.received_flag[chunk_number % self.buffer_size]:
            sys.stdout.write(Color.cyan)
            _print_("lost chunk", chunk_number)
            sys.stdout.write(Color.none)
            chunk_number = (chunk_number + 1) % common.MAX_CHUNK_NUMBER
            #counter += 1
            #if counter > self.buffer_size:
            #    break
        return chunk_number

        # }}}

    def play_chunk(self, chunk):
        # {{{

        try:
            self.player_socket.sendall(self.chunks[chunk % self.buffer_size])
        except socket.error:
            #print(e)
            _print_("Player disconnected!")
            self.player_alive = False

        # }}}

    def play_next_chunk(self):
        # {{{

        self.played_chunk = self.find_next_chunk()
        self.play_chunk(self.played_chunk)
        self.received_flag[self.played_chunk % self.buffer_size] = False
        self.received_counter -= 1
        #print("----------------------------")

        # }}}

    def play(self):
        # {{{
        
        while self.player_alive:
            #self.keep_the_buffer_full()
            self.play_next_chunk()

        # }}}

    ## def receive_configuration(self):
    ##     # {{{

    ##     self.receive_the_mcast_endpoint()
    ##     self.receive_the_header_size()
    ##     self.receive_the_chunk_size()
    ##     self.receive_the_header()
    ##     self.receive_the_buffer_size()

    ##     # }}}
        
    def keep_the_buffer_full(self):
        # {{{

        # Receive chunks while the buffer is not full
        #while True:
        #    chunk_number = self.process_next_message()
        #    if chunk_number >= 0:
        #        break
        chunk_number = self.process_next_message()
        while chunk_number < 0:
            chunk_number = self.process_next_message()
        #while ((chunk_number - self.played_chunk) % self.buffer_size) < self.buffer_size/2:
        while self.received_counter < self.buffer_size/2:
            chunk_number = self.process_next_message()
            while chunk_number < 0:
                chunk_number = self.process_next_message()

        if __debug__:
            for i in range(self.buffer_size):
                if self.received_flag[i]:
                    sys.stdout.write(str(i%10))
                else:
                    sys.stdout.write('.')
            print ()
            #print (self.team_socket.getsockname(),)
            sys.stdout.write(Color.none)

        # }}}

    def run(self):
        # {{{

        #threading.Thread(target=self.play).start()

        while self.player_alive:
            self.keep_the_buffer_full()
            self.play_next_chunk()

        # }}}

    # }}}

File: src/core/peer_ims.cpp

#include "peer_ims.h"

namespace p2psp {

constexpr char PeerIMS::kSplitterAddr[];

PeerIMS::PeerIMS()
    : io_service_(),
      acceptor_(io_service_),
      player_socket_(io_service_),
      splitter_socket_(io_service_),
      team_socket_(io_service_) {
  // Default values
  player_port_ = kPlayerPort;
  splitter_addr_ = ip::address::from_string(kSplitterAddr);
  splitter_port_ = kSplitterPort;
  port_ = kPort;
  use_localhost_ = kUseLocalhost;
  buffer_status_ = kBufferStatus;
  show_buffer_ = kShowBuffer;

  // Initialized in PeerIMS::ReceiveTheBufferSize()
  buffer_size_ = 0;

  // Initialized in PeerIMS::ReceiveTheChunkSize()
  chunk_size_ = 0;
  chunks_ = std::vector<Chunk>();

  // Initialized in PeerIMS::ReceiveTheHeaderSize()
  header_size_in_chunks_ = 0;

  // Initialized in PeerIMS::ReceiveTheMcasteEndpoint()
  mcast_addr_ = ip::address::from_string("0.0.0.0");
  mcast_port_ = 0;

  played_chunk_ = 0;
  player_alive_ = false;

  received_counter_ = 0;
  received_flag_ = std::vector<bool>();
  recvfrom_counter_ = 0;

  sendto_counter_ = -1;
}

PeerIMS::~PeerIMS() {}

void PeerIMS::Init(){};

void PeerIMS::WaitForThePlayer() {
  std::string port = std::to_string(player_port_);
  ip::tcp::resolver resolver(io_service_);
  ip::tcp::endpoint endpoint = *resolver.resolve({"", port});

  acceptor_.open(endpoint.protocol());
  acceptor_.set_option(ip::tcp::acceptor::reuse_address(true));
  acceptor_.bind(endpoint);
  acceptor_.listen();

  LOG("Waiting for the player at (" << endpoint.address().to_string() << ","
                                    << std::to_string(endpoint.port()) << ")");
  acceptor_.accept(player_socket_);

  LOG("The player is ("
      << player_socket_.remote_endpoint().address().to_string() << ","
      << std::to_string(player_socket_.remote_endpoint().port()) << ")");
}

void PeerIMS::ConnectToTheSplitter() {
  std::string my_ip;

  // TCP endpoint object to connect to splitter
  ip::tcp::endpoint splitter_tcp_endpoint(splitter_addr_, splitter_port_);
  // UDP endpoint object to connect to splitter
  splitter_ = ip::udp::endpoint(splitter_addr_, splitter_port_);

  ip::tcp::endpoint tcp_endpoint;

  LOG("use_localhost = " << std::string((use_localhost_ ? "True" : "False")));
  if (use_localhost_) {
    my_ip = "0.0.0.0";
  } else {
    ip::udp::socket s(io_service_);
    try {
      s.connect(splitter_);
    } catch (boost::system::system_error e) {
      LOG(e.what());
    }

    my_ip = s.local_endpoint().address().to_string();
    s.close();
  }

  splitter_socket_.open(splitter_tcp_endpoint.protocol());

  LOG("Connecting to the splitter at ("
      << splitter_tcp_endpoint.address().to_string() << ","
      << std::to_string(splitter_tcp_endpoint.port()) << ") from " << my_ip);
  if (port_ != 0) {
    LOG("I'm using port" << std::to_string(port_));
    tcp_endpoint = ip::tcp::endpoint(ip::address::from_string(my_ip), port_);
    splitter_socket_.set_option(ip::udp::socket::reuse_address(true));
  } else {
    tcp_endpoint = ip::tcp::endpoint(ip::address::from_string(my_ip), 0);
  }

  splitter_socket_.bind(tcp_endpoint);

  try {
    splitter_socket_.connect(splitter_tcp_endpoint);
  } catch (boost::system::system_error e) {
    if (IFF_DEBUG) {
      LOG(e.what());
    } else {
      LOG(e.what());
    }
    exit(-1);
  }

  LOG("Connected to the splitter at ("
      << splitter_tcp_endpoint.address().to_string() << ","
      << std::to_string(splitter_tcp_endpoint.port()) << ")");
}

void PeerIMS::DisconnectFromTheSplitter() { splitter_socket_.close(); }

void PeerIMS::ReceiveTheMcasteEndpoint() {
  boost::array<char, 6> buffer;
  read(splitter_socket_, ::buffer(buffer));

  char *raw_data = buffer.data();

  in_addr ip_raw = *(in_addr *)(raw_data);
  mcast_addr_ = ip::address::from_string(inet_ntoa(ip_raw));
  mcast_port_ = ntohs(*(short *)(raw_data + 4));

  LOG("mcast_endpoint = (" << mcast_addr_.to_string() << ","
                           << std::to_string(mcast_port_) << ")");
}

void PeerIMS::ReceiveTheHeaderSize() {
  boost::array<char, 2> buffer;
  read(splitter_socket_, ::buffer(buffer));

  header_size_in_chunks_ = ntohs(*(short *)(buffer.c_array()));

  LOG("header_size (in chunks) = " << std::to_string(header_size_in_chunks_));
}

void PeerIMS::ReceiveTheChunkSize() {
  boost::array<char, 2> buffer;
  read(splitter_socket_, ::buffer(buffer));

  chunk_size_ = ntohs(*(short *)(buffer.c_array()));

  LOG("chunk_size (bytes) = " << std::to_string(chunk_size_));
}

void PeerIMS::ReceiveTheHeader() {
  int header_size_in_bytes = header_size_in_chunks_ * chunk_size_;
  std::vector<char> header(header_size_in_bytes);

  boost::system::error_code ec;
  streambuf chunk;

  read(splitter_socket_, chunk, transfer_exactly(header_size_in_bytes), ec);
  if (ec) {
    LOG("Error: " << ec.message());
  }

  try {
    write(player_socket_, chunk);
  } catch (std::exception e) {
    LOG(e.what());
    LOG("error sending data to the player");
    LOG("len(data) =" << std::to_string(chunk.size()));
    boost::this_thread::sleep(boost::posix_time::milliseconds(1000));
  }

  LOG("Received " << std::to_string(header_size_in_bytes) << "bytes of header");
}

void PeerIMS::ReceiveTheBufferSize() {
  boost::array<char, 2> buffer;
  read(splitter_socket_, ::buffer(buffer));

  buffer_size_ = ntohs(*(short *)(buffer.c_array()));

  LOG("buffer_size_ = " << std::to_string(buffer_size_));
}

void PeerIMS::ListenToTheTeam() {
  ip::udp::endpoint endpoint(ip::address_v4::any(), mcast_port_);

  team_socket_.open(endpoint.protocol());
  team_socket_.set_option(ip::udp::socket::reuse_address(true));
  team_socket_.bind(endpoint);

  team_socket_.set_option(ip::multicast::join_group(mcast_addr_));

  // TODO: handle timeout
  LOG("Listening to the mcast_channel = ("
      << mcast_addr_.to_string() << "," << std::to_string(mcast_port_) << ")");
}

void PeerIMS::BufferData() {
  // The peer dies if the player disconnects.
  player_alive_ = true;

  // The last chunk sent to the player.
  played_chunk_ = 0;

  // Counts the number of executions of the recvfrom() function.
  recvfrom_counter_ = 0;

  // The buffer of chunks is a structure that is used to delay the playback of
  // the chunks in order to accommodate the network jittter. Two components are
  // needed: (1) the "chunks" buffer that stores the received chunks and (2) the
  // "received" buffer that stores if a chunk has been received or not. Notice
  // that each peer can use a different buffer_size: the smaller the buffer
  // size, the lower start-up time, the higher chunk-loss ratio. However, for
  // the sake of simpliticy, all peers will use the same buffer size.

  chunks_.resize(buffer_size_);
  received_counter_ = 0;

  // Wall time (execution time plus waiting time).
  clock_t start_time = clock();

  // We will send a chunk to the player when a new chunk is received. Besides,
  // those slots in the buffer that have not been filled by a new chunk will not
  // be send to the player. Moreover, chunks can be delayed an unknown time.
  // This means that (due to the jitter) after chunk X, the chunk X+Y can be
  // received (instead of the chunk X+1). Alike, the chunk X-Y could follow the
  // chunk X. Because we implement the buffer as a circular queue, in order to
  // minimize the probability of a delayed chunk overwrites a new chunk that is
  // waiting for traveling the player, we wil fill only the half of the circular
  // queue.

  LOG("(" << team_socket_.local_endpoint().address().to_string() << ","
          << std::to_string(team_socket_.local_endpoint().port()) << ")"
          << "\b: buffering = 000.00%");
  TraceSystem::logStream().flush();

  // First chunk to be sent to the player.  The process_next_message() procedure
  // returns the chunk number if a packet has been received or -2 if a time-out
  // exception has been arised.

  int chunk_number = ProcessNextMessage();
  while (chunk_number < 0) {
    chunk_number = ProcessNextMessage();
    LOG(std::to_string(chunk_number));
  }
  played_chunk_ = chunk_number;
  LOG("First chunk to play " << std::to_string(played_chunk_));
  LOG("(" << team_socket_.local_endpoint().address().to_string() << ","
          << std::to_string(team_socket_.local_endpoint().port()) << ")"
          << "\b: buffering (\b" << std::to_string(100.0 / buffer_size_));
  // TODO: Justify: .rjust(4)

  // Now, fill up to the half of the buffer.

  float BUFFER_STATUS = 0.0f;
  for (int x = 0; x < buffer_size_ / 2; x++) {
    // TODO Format string
    // LOG("{:.2%}\r".format((1.0*x)/(buffer_size_/2)), end='');
    BUFFER_STATUS = (100 * x) / (buffer_size_ / 2.0f) + 1;

    if (!Common::kConsoleMode) {
      // GObject.idle_add(buffering_adapter.update_widget,BUFFER_STATUS)
    } else {
      // pass
    }
    LOG("!");
    TraceSystem::logStream().flush();

    while (ProcessNextMessage() < 0)
      ;
  }

  LOG("");
  LOG("latency = " << std::to_string((clock() - start_time) /
                                     (float)CLOCKS_PER_SEC) << " seconds");
  LOG("buffering done.");
  TraceSystem::logStream().flush();
}

int PeerIMS::ProcessNextMessage() {
  // (Chunk number + chunk payload) length
  std::vector<char> message(sizeof(uint16_t) + chunk_size_);
  ip::udp::endpoint sender;

  try {
    ReceiveTheNextMessage(&message, sender);
  } catch (std::exception e) {
    return -2;
  }

  return ProcessMessage(message, sender);
}

void PeerIMS::ReceiveTheNextMessage(std::vector<char> *message,
                                    ip::udp::endpoint sender) {
  LOG("Waiting for a chunk at ("
      << team_socket_.local_endpoint().address().to_string() << ","
      << std::to_string(team_socket_.local_endpoint().port()) << ")");

  team_socket_.receive_from(buffer(*message), sender);
  recvfrom_counter_++;

  LOG("Received a message from ("
      << sender.address().to_string() << "," << std::to_string(sender.port())
      << ") of length " << std::to_string(message->size()));

  // TODO: if(DEBUG){
  if (message->size() < 10) {
    LOG("Message content =" << std::string(message->data()));
  }
  //}
}

int PeerIMS::ProcessMessage(std::vector<char> message,
                            ip::udp::endpoint sender) {
  // Ojo, an attacker could send a packet smaller and pollute the buffer,
  // althought this is difficult in IP multicst. This method should be
  // inheritaged to solve this issue.

  uint16_t chunk_number = ntohs(*(short *)message.data());

  chunks_[chunk_number % buffer_size_] = {
      std::vector<char>(message.data() + sizeof(uint16_t),
                        message.data() + message.size()),
      true};

  received_counter_++;

  return chunk_number;
}

void PeerIMS::KeepTheBufferFull() {
  // Receive chunks while the buffer is not full
  // while True:
  //    chunk_number = self.process_next_message()
  //    if chunk_number >= 0:
  //        break

  int chunk_number = ProcessNextMessage();
  while (chunk_number < 0) {
    chunk_number = ProcessNextMessage();
  }
  // while ((chunk_number - self.played_chunk) % self.buffer_size) <
  // self.buffer_size/2:
  while (received_counter_ < buffer_size_ / 2) {
    chunk_number = ProcessNextMessage();
    while (chunk_number < 0) {
      chunk_number = ProcessNextMessage();
    }
  }

  if (show_buffer_) {
    for (int i = 0; buffer_size_; i++) {
      if (chunks_[i].received) {
        // TODO: Avoid line feed in LOG function
        LOG(std::to_string(i % 10));
      } else {
        LOG(".");
      }
    }
    LOG("");
  }

  // print (self.team_socket.getsockname(),)
  // sys.stdout.write(Color.none)
}

void PeerIMS::PlayNextChunk() {
  played_chunk_ = FindNextChunk();
  PlayChunk(played_chunk_);
  chunks_[played_chunk_].received = false;
  received_counter_--;
}

// Tiene pinta de que los tres siguientes metodos pueden simplificarse...
int PeerIMS::FindNextChunk() {
  // print (".")
  // counter = 0

  int chunk_number = (played_chunk_ + 1) % Common::kMaxChunkNumber;

  while (!chunks_[chunk_number % buffer_size_].received) {
    // sys.stdout.write(Color.cyan)
    LOG("lost chunk " << std::to_string(chunk_number));
    // sys.stdout.write(Color.none)

    chunk_number = (played_chunk_ + 1) % Common::kMaxChunkNumber;
  }
  // counter++
  // if counter > self.buffer_size:
  //    break
  return chunk_number;
}

void PeerIMS::PlayChunk(int chunk) {
  try {
    write(player_socket_, buffer(chunks_[chunk % buffer_size_].data));
  } catch (std::exception e) {
    LOG("Player disconnected!");
    player_alive_ = false;
  }
}

void PeerIMS::Run() {
  while (player_alive_) {
    KeepTheBufferFull();
    PlayNextChunk();
  }
}

void PeerIMS::Start() {
  thread_.reset(new boost::thread(boost::bind(&PeerIMS::Run, this)));
}

std::string PeerIMS::GetMcastAddr() { return mcast_addr_.to_string(); }

void PeerIMS::SetShowBuffer(bool show_buffer) { show_buffer_ = show_buffer; }

bool PeerIMS::IsPlayerAlive() { return player_alive_; }

int PeerIMS::GetPlayedChunk() { return played_chunk_; }

int PeerIMS::GetChunkSize() { return chunk_size_; }

int PeerIMS::GetRecvfromCounter() { return recvfrom_counter_; }

std::vector<ip::udp::endpoint> *PeerIMS::GetPeerList() { return &peer_list_; }

int PeerIMS::GetSendtoCounter() { return sendto_counter_; }
void PeerIMS::SetSendtoCounter(int sendto_counter) {
  sendto_counter_ = sendto_counter;
}
}